### ISSUE TITLE
Raylet Memory  usage growing 

### ISSUE THREAD

**Issue by KemalAcar07** [ray-project/ray#4657]
### System information
- **OS Platform and Distribution:**  Ubuntu 16.04
- **Ray installed from:** binary
- **Ray version**: 0.6.5
- **Python version**:3.6.8
- **Exact command to reproduce**

```python
import numpy as np
import ray
import time
ray.init(redis_max_memory=100000000)


@ray.remote
class Runner():
    def __init__(self, dataList):
        self.run(dataList)

    def run(self,dataList):
        while True:
            dataList.put.remote(np.ones(10))

@ray.remote
class Optimizer():
    def __init__(self, dataList):
        self.optimize(dataList)

    def optimize(self,dataList):
        while True:
            dataList.pop.remote()

@ray.remote
class DataServer():
    def __init__(self):
        self.dataList= []

    def put(self,data):
        self.dataList.append(data)

    def pop(self):
        if len(self.dataList) !=0:
            return self.dataList.pop()
    def get_size(self):
        return len(self.dataList)


dataServer = DataServer.remote()
runner = Runner.remote(dataServer)
optimizer1 = Optimizer.remote(dataServer)
optimizer2 = Optimizer.remote(dataServer)

while True:
    time.sleep(1)
    print(ray.get(dataServer.get_size.remote()))
```

### Describe the problem
The memory usage is constantly increasing even though the list is cleared repeatedly until the program crashes. Does anybody know why this is happening ?

### Source code / logs

Traceback (most recent call last):
  File "/home/test.py", line 48, in <module>
    
  File "/home/anaconda3/envs/py3/lib/python3.6/site-packages/ray/worker.py", line 2310, in get
    value = worker.get_object([object_ids])[0]
  File "/home/anaconda3/envs/py3/lib/python3.6/site-packages/ray/worker.py", line 531, in get_object
    self.current_task_id,
  File "python/ray/_raylet.pyx", line 254, in ray._raylet.RayletClient.fetch_or_reconstruct
  File "python/ray/_raylet.pyx", line 59, in ray._raylet.check_status
Exception: [RayletClient] Connection closed unexpectedly.

Process finished with exit code 1

**Comment by ericl** [comment]
Can you provide more details on what sort of memory usage you are seeing, and where it is coming from? For example, memory keeps increasing from 10 to 30 GB and my node only has 30 GB.

Btw in your example you didn't set an object store max size, the default is 30% of memory.

**Comment by KemalAcar07** [comment]
I don't think that the object store is the problem here.
The memory usage on this particular computer is before starting this code sample at 4.39 GB out of16 GB and when I start it it grows approximately 0.15 GB per second  until the maximum capacity is reached. After that the above listed error message appears. The process name that is taking all the memory is called raylet. The object store process is only taking 0.75 GB at the end of the execution.

**Comment by ericl** [comment]
Thanks @KemalAcar07 , I can confirm the result. I see the `raylet` process also increasing in memory, though the other processes are stable in memory usage.

At first glance, this may be related to this: https://github.com/ray-project/ray/issues/4359, but will require more investigation. cc @robertnishihara

**Comment by ericl** [comment]
It seems unrelated to #4359 which is now fixed. Actually, I ran this workload on master and now it crashes with 
```
WARNING: Logging before InitGoogleLogging() is written to STDERR
I0610 15:29:15.357662  2229 stats.h:48] Succeeded to initialize stats: exporter address is 127.0.0.1:8888
F0610 15:29:47.084838  2229 scheduling_resources.cc:182]  Check failed: resource_capacity_[resource_label] >= 0 Capacity of resource after subtraction is negative, -214747.
*** Check failure stack trace: ***
*** Aborted at 1560205787 (unix time) try "date -d @1560205787" if you are using GNU date ***
PC: @                0x0 (unknown)
*** SIGABRT (@0x3e8000008b5) received by PID 2229 (TID 0x7f4f354c9740) from PID 2229; stack trace: ***
    @     0x7f4f350c2390 (unknown)
    @     0x7f4f34273428 gsignal
    @     0x7f4f3427502a abort
    @           0x501d09 google::logging_fail()
    @           0x503b8a google::LogMessage::Fail()
    @           0x504f6a google::LogMessage::SendToLog()
    @           0x5038b5 google::LogMessage::Flush()
    @           0x503ab1 google::LogMessage::~LogMessage()
    @           0x501251 ray::RayLog::~RayLog()
    @           0x450bdc ray::raylet::ResourceSet::SubtractResourcesStrict()
    @           0x453530 ray::raylet::TaskQueue::RemoveTask()
    @           0x456790 ray::raylet::SchedulingQueue::RemoveTasksFromQueue()
    @           0x456bc4 ray::raylet::SchedulingQueue::RemoveTasks()
    @           0x46df34 ray::raylet::NodeManager::HandleObjectLocal()
    @           0x474a2e ray::raylet::NodeManager::FinishAssignedActorTask()
    @           0x4752c6 ray::raylet::NodeManager::FinishAssignedTask()
    @           0x475419 ray::raylet::NodeManager::ProcessGetTaskMessage()
    @           0x475d8b ray::raylet::NodeManager::ProcessClientMessage()
    @           0x433880 _ZNSt17_Function_handlerIFvSt10shared_ptrIN3ray16ClientConnectionIN5boost4asio5local15stream_protocolEEEElPKhEZNS1_6raylet6Raylet12HandleAcceptERKNS3_6system10error_codeEEUlS8_lSA_E0_E9_M_invokeERKSt9_Any_dataOS8_OlOSA_
    @           0x4fbaf3 ray::ClientConnection<>::ProcessMessage()
    @           0x4f810c boost::asio::detail::read_op<>::operator()()
    @           0x4f82af boost::asio::detail::reactive_socket_recv_op<>::do_complete()
    @           0x42a48c boost::asio::detail::scheduler::run()
    @           0x410f4e main
    @     0x7f4f3425e830 __libc_start_main
    @           0x422ed9 _start
```

**Comment by ericl** [comment]
Upon further inspection, this is actually due to an unbounded number of task buildup in the scheduler. This is because the example submits tasks in a while: loop without calling ray.get, so there is no backpressure.

Changing the blocks from:
```
        while True:
            dataList.put.remote(np.ones(10))
```
to
```
        while True:
            ray.get(dataList.put.remote(np.ones(10)))
```

causes memory to not build up at all.

The crash I posted above is actually a separate issue, which also happens to be resolved if the above fix is made (hmm...).

**Comment by simon-mo** [comment]
Should be fixed by direct task call and direct actor call re-architecture. please reopen if this still fails on latest master/release.
